### DISCUSSION TITLE
Re-inserts of deleted records, records reverted to older states unintentionally

### DISCUSSION THREAD

**Discussion by scottboyd-ai** [mikro-orm/mikro-orm#3232] [Q&A]
**Describe the bug**
This bug is difficult to describe, as it seems somewhat complicated. I have an Events table with a Sign_Up table that contains a foreign key reference to Events. The Event reference on Sign_Up has the `onDelete: 'cascade'` option turned on. When I delete an event using `DI.em.removeAndFlush()`, the transaction completes correctly. However, when I complete another transaction that selects all records from the Sign_Up table, the Event record that I just deleted is re-inserted. The Sign_Up records are still deleted (not re-inserted). With logging and debugging, I've discovered that this happens during a `DI.em.find()` call, or during a `flush()` call.

I am on MikroORM V5, and `DI.em` is a global variable defined with the following block:

```
DI.orm = await MikroORM.init();
DI.em = DI.orm.em.fork({freshEventManager: true});
```

`DI` is a const exported in Server.ts, where the main startup logic is also written. DI is defined as follows:

```
export const DI = {} as {
    orm: MikroORM<PostgreSqlDriver>,
    em: EntityManager,
    qb: QueryBuilder
};
```

To me, it seems like MikroORM is keeping the reference to this record in memory, and when it is referenced but not present in the DB, it re-inserts the record to keep things clean.

I have worked around this particular issue by explicitly deleting the Sign_Up records, then explicitly deleting the Events record.

I'm having a similar, possibly related issue where a record was somehow restored to an older state without any user interaction. A record on the Event table was somehow updated back to a state it had been a few days prior.

Here is my Event entity: (Base model is simply an ID field, uuidv4)

```
@Entity()
export class Event extends BaseModel{

    @Property()
    title: string;

    @Property({nullable: true, columnType: 'text'})
    description?: string;

    @Property({type: NativeBigIntType})
    startDate: number;

    @Property({nullable: true, type: NativeBigIntType})
    endDate?: number;

    @ManyToOne( {nullable: true, onDelete: "SET NULL"})
    type?: EventType;

    [key: string]: any;

}
```

My Sign_Up table is defined as follows:

```
@Entity()
export class SignUp extends BaseModel {

    @ManyToOne()
    user: User;

    @ManyToOne({onDelete: 'cascade'})
    event: Event;

    @Property({type: NativeBigIntType})
    eventStartDate: number;

    @Property({nullable: true, type: NativeBigIntType})
    eventEndDate?: number;

    @Property({default: false})
    isCanceled: boolean = false;

    @Property({nullable: true, type: NativeBigIntType})
    clockInTime?: number;

    @Property({nullable: true, type: NativeBigIntType})
    clockOutTime?: number;

}
```

**Stack trace**
No stacktrace, but I do see the sql statements for each of these events. They are completely normal statements, but unexpected and unwanted.

**To Reproduce**
1. Create an Event record
2. Create a Sign_Up record that references the Event.
3. 
```
const eventToDelete = await DI.em.findOne(Event, {id: eventId})
if (eventToDelete) {
        await DI.em.removeAndFlush(eventToDelete);
}

```
4. 
```
const events = await DI.em.find(Event, {}, {populate: ['type']});
```

**Expected behavior**
The ORM returns a list of all events and does nothing else.

**Additional context**
I'm happy to provide any additional information needed. I am running MikroORM as a connection to a PostgreSQL database. I am using Express as a server framework.

**Versions**

| Dependency | Version |
| - | - |
| node | 14.15.4 |
| typescript | 4.5.5 |
| mikro-orm | 5.1.5 |
| postgresql-driver | 5.1.5 |

**Comment by B4nan** [comment]
`onDelete: 'cascade'` is a database level option, it can't affect how the UoW behaves. What is probably happening to you is that you keep the entities in your identity map, let the real records be removed by referential integrity (via database directly), but dont clean up the identity map properly. It should happen automatically if you have cascading set up correctly. 

I can't help more than this without a complete reproduction, what you provided is not enough to reproduce it. Note that the reproduction should be ideally a single script, there is no need for handling dependency injection or any web servers like express, no additional dependencies needed basically. Also always upgrade to latest before reporting anything.

In recent versions we started to respect the `onDelete` option too for UoW, maybe we could deal with this automatically, but I will need to see complete reproduction first to be able to tell that. Because it might as well be a wrongly managed request context or something like that.

**Reply by scottboyd-ai** [reply]
That was my initial thought as well, but I was under the impression that having the global `DI.em` be defined as `DI.em.fork()` would have a unique identity map each time it was called. Do I need to fork on each API route, or each time I want a new identity map? Or should I be switching the `onDelete: 'cascade'` to `cascade: [Cascade.ALL]`? The distinction isn't very clear.

**Reply by B4nan** [reply]
Oh I missed that. That's wrong for sure, or rather a no-op. You just replace a global instance with another (still global) instance, all what that will help you with is disabling the global context validation. Forking needs to happen on a request basis, not once on initialization.
